content/vault/v1.18.x/content/docs/upgrading/upgrade-to-1.16.x.mdx

```diff
@@ -18,6 +18,48 @@ Vault 1.15. **Please read carefully**.
 
 ## Important changes
 
+### JSON Payload Limits
+
+| Change       | Affected version | Vault edition
+| ------------ | ---------------- | -------------
+| New behavior | 1.16.25           | All
+|              |                  |              |
+|              |                  |              |
+
+To guard against potential Denial-of-Service (DoS) attacks, this release
+introduces several Vault listener options that enforce payload size limits applied to incoming JSON request bodies.
+These settings are configured on a listener and give administrators granular control over the
+structure and size of JSON payloads that Vault will process and the configuration and performance characteristics of their Vault cluster.
+We've chosen defaults that are intentionally generous to accommodate a wide range of
+legitimate use cases, and which should guard against most malicious or malformed requests.
+
+`max_json_depth`
+This parameter specifies the maximum allowed nesting depth of a JSON object or array.
+It serves as a critical defense against "JSON depth bomb" attacks, where a malicious
+client sends a deeply nested payload to cause a stack overflow on the server.
+The default value is 500, which is well above the nesting level of any typical
+Vault configuration or secret, providing a robust safeguard without impacting normal operations.
+
+`max_json_string_value_length`
+This parameter defines the maximum allowed length for any single string value
+within a JSON payload, measured in bytes. This limit prevents a client from
+sending a request with an extremely large string to cause excessive memory allocation on the server.
+The default value is 1,048,576 bytes (1MB), which is aligned with Vault's default storage backend
+limit and is large enough to comfortably accommodate common secrets like full TLS certificate chains.
+
+`max_json_object_entry_count`
+This parameter sets the maximum number of key-value pairs allowed in a single JSON object.
+This is a direct defense against HashDoS (Hash Flooding) attacks, where an attacker could send
+an object with an enormous number of keys to overwhelm the server's CPU. A legitimate JSON object
+with more than 10,000 keys is extremely rare, so the default of 10,000 provides a high ceiling for normal operations
+while ensuring server stability.
+
+`max_json_array_element_count`
+This parameter determines the maximum number of elements permitted in a single JSON array.
+Its primary purpose is to protect the server from having to construct and hold a massive,
+memory-intensive array in a single API response, which is most common during LIST operations on secrets paths.
+The default of 10,000 elements acts as a sensible guardrail.
```

Comment:
```suggestion
```

The important changes page isn't really the place to document the details of each parameter, we should provide a one-sentence summary for each and link out to the relevant config docs for more information